### PR TITLE
Introduce dynamic hook loading

### PR DESCRIPTION
**Pull request by nupplaphil**
- Dynamically load addon files
- Dynamically load hooks
- Rewrite Logger-logic to use new hook logic (Monolog is working again)

Monolog fix: https://git.friendi.ca/friendica/friendica-addons/pulls/1400 

ToDo
- [x] Add as much phpunit-tests as possible :)
- [x] Test on friendica.philipp.info

I'm now finished. It's running on friendica.philipp.info

I tried to make parts of the logic easier in removing the "Decorator" hook logic. It seems like they're just used for edge cases, which can be covered directly in factories (Like for the `ProfilerLogger`).

I already prepared a followup for pipelining & event hook logic, which will cover all other types of hooks so far.

This hook logic can be used everywhere in the codebase to decouple namespaces .. It would be at least an entry point for "real" Domain Driven Design, which I'm no more forcing, but trying to add as much potential entrypoints as possible ;-)

[Edit] Btw. the woodpecker CI instance isn't working properly. I think I will have some time in the next 2 weeks to hopefully fix it...

### CODE REVIEW

File: bin/daemon.php [hunk 1]
```diff
@@ -60,7 +60,10 @@
 require dirname(__DIR__) . '/vendor/autoload.php';
 
 $dice = (new Dice())->addRules(include __DIR__ . '/../static/dependencies.config.php');
-$dice = $dice->addRule(LoggerInterface::class,['constructParams' => ['daemon']]);
+/** @var \Friendica\Core\Addon\Capabilities\ICanLoadAddons $addonLoader */
+$addonLoader = $dice->create(\Friendica\Core\Addon\Capabilities\ICanLoadAddons::class);
+$dice = $dice->addRules($addonLoader->getActiveAddonConfig('dependencies'));
+$dice = $dice->addRule(LoggerInterface::class,['constructParams' => [Logger\Capabilities\LogChannel::DAEMON]]);
```
Comment (MrPetovan) [inline bin/daemon.php]:
```suggestion
$dice = $dice->addRule(LoggerInterface::class, ['constructParams' => [Logger\Capabilities\LogChannel::DAEMON]]);
```

File: bin/worker.php [hunk 2]
```diff
@@ -54,7 +55,10 @@
 require dirname(__DIR__) . '/vendor/autoload.php';
 
 $dice = (new Dice())->addRules(include __DIR__ . '/../static/dependencies.config.php');
-$dice = $dice->addRule(LoggerInterface::class,['constructParams' => ['worker']]);
+/** @var \Friendica\Core\Addon\Capabilities\ICanLoadAddons $addonLoader */
+$addonLoader = $dice->create(\Friendica\Core\Addon\Capabilities\ICanLoadAddons::class);
+$dice = $dice->addRules($addonLoader->getActiveAddonConfig('dependencies'));
+$dice = $dice->addRule(LoggerInterface::class,['constructParams' => [LogChannel::WORKER]]);
```
Comment (MrPetovan) [inline bin/worker.php]:
```suggestion
$dice = $dice->addRule(LoggerInterface::class, ['constructParams' => [LogChannel::WORKER]]);
```

File: src/Core/Addon/Exception/AddonInvalidConfigFileException.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+namespace Friendica\Core\Addon\Exception;
+
+use Throwable;
+
+/**
+ * Exception in case one or more config files of the addons are invalid
+ */
+class AddonInvalidConfigFileException extends \RuntimeException
+{
+	public function __construct($message = "", $code = 0, Throwable $previous = null)
```
Comment (MrPetovan) [inline src/Core/Addon/Exception/AddonInvalidConfigFileException.php]:
```suggestion
	public function __construct($message = '', $code = 0, Throwable $previous = null)
```

File: src/Core/Hooks/Capabilities/ICanCreateInstances.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	 *
+	 * The instance will be build based on the registered strategy and the (unique) name
+	 *
+	 * @param string $class     The fully-qualified name of the given class or interface which will get returned
+	 * @param string $name      An arbitrary identifier to find a concrete instance strategy.
+	 * @param array  $arguments Additional arguments, which can be passed to the constructor of "$class" at runtime
+	 *
+	 * @return object The concrete instance of the type "$class"
+	 */
+	public function create(string $class, string $name, array $arguments = []): object;
```
Comment (MrPetovan) [inline src/Core/Hooks/Capabilities/ICanCreateInstances.php]:
```suggestion
	public function create(string $class, string $strategy, array $arguments = []): object;
```

File: src/Core/Hooks/Capabilities/ICanRegisterInstances.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+namespace Friendica\Core\Hooks\Capabilities;
+
+use Friendica\Core\Hooks\Exceptions\HookRegisterArgumentException;
+
+/**
+ * Register strategies for given classes
+ */
+interface ICanRegisterInstances
```
Comment (MrPetovan) [inline src/Core/Hooks/Capabilities/ICanRegisterInstances.php]: Shouldn't it be `ICanRegisterStrategy`?

File: src/Core/Hooks/Capabilities/ICanRegisterInstances.php [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+{
+	/**
+	 * Register a class(strategy) for a given interface with a unique name.
+	 *
+	 * @see https://refactoring.guru/design-patterns/strategy
+	 *
+	 * @param string  $interface The interface, which the given class implements
+	 * @param string  $class     The fully-qualified given class name
+	 *                           A placeholder for dependencies is possible as well
+	 * @param ?string $name      An arbitrary identifier for the given class, which will be used for factories, dependency injections etc.
```
Comment (MrPetovan) [inline src/Core/Hooks/Capabilities/ICanRegisterInstances.php]: For the given class or the given strategy?

File: src/Core/Hooks/Exceptions/HookConfigException.php [hunk 7]
```diff
@@ -19,11 +19,12 @@
...
-namespace Friendica\Core\Hooks\Capabilities;
+namespace Friendica\Core\Hooks\Exceptions;
 
-/**
- * All classes, implementing this interface are valid Strategies for Hook calls
- */
-interface IAmAStrategy
+class HookConfigException extends \RuntimeException
 {
+	public function __construct($message = "", \Throwable $previous = null)
```
Comment (MrPetovan) [inline src/Core/Hooks/Exceptions/HookConfigException.php]:
```suggestion
	public function __construct($message = '', \Throwable $previous = null)
```

File: src/Core/Hooks/Util/HookFileManager.php [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+	 *
+	 * @param ICanRegisterInstances $instanceRegister The instance register
+	 *
+	 * @return void
+	 */
+	public function setupHooks(ICanRegisterInstances $instanceRegister)
+	{
+		// In case it wasn't used before, reload the whole hook config
+		if (empty($this->hookConfig)) {
+			$this->reloadHookConfig();
```
Comment (MrPetovan) [inline src/Core/Hooks/Util/HookFileManager.php]: Shouldn't this be done in the constructor?

File: src/Core/Logger/Exception/LogLevelException.php [hunk 9]
```diff
@@ -23,6 +23,9 @@
 
 use Throwable;
 
+/**
+ * Exception in case the loglevel isn't set or isn't valid
+ */
 class LogLevelException extends \InvalidArgumentException
 {
 	public function __construct($message = "", Throwable $previous = null)
```
Comment (MrPetovan) [inline src/Core/Logger/Exception/LogLevelException.php]:
```suggestion
	public function __construct($message = '', Throwable $previous = null)
```

File: src/Core/Logger/Exception/LoggerArgumentException.php [hunk 10]
```diff
@@ -23,6 +23,9 @@
 
 use Throwable;
 
+/**
+ * Exception in case an argument of a logger class isn't valid
+ */
 class LoggerArgumentException extends \InvalidArgumentException
 {
 	public function __construct($message = "", Throwable $previous = null)
```
Comment (MrPetovan) [inline src/Core/Logger/Exception/LoggerArgumentException.php]:
```suggestion
	public function __construct($message = '', Throwable $previous = null)
```

File: src/Core/Logger/Exception/LoggerException.php [hunk 11]
```diff
@@ -23,6 +23,9 @@
 
 use Throwable;
 
+/**
+ * A generic exception of the logging namespace
+ */
 class LoggerException extends \Exception
 {
 	public function __construct($message = "", Throwable $previous = null)
```
Comment (MrPetovan) [inline src/Core/Logger/Exception/LoggerException.php]:
```suggestion
	public function __construct($message = '', Throwable $previous = null)
```

File: src/Core/Logger/Exception/LoggerUnusableException.php [hunk 12]
```diff
@@ -23,7 +23,10 @@
 
 use Throwable;
 
-class LoggerInvalidException extends \RuntimeException
+/**
+ * Exception in case the used logging instance is unusable because of some circumstances
+ */
+class LoggerUnusableException extends \RuntimeException
 {
 	public function __construct($message = "", Throwable $previous = null)
```
Comment (MrPetovan) [inline src/Core/Logger/Exception/LoggerUnusableException.php]:
```suggestion
	public function __construct($message = '', Throwable $previous = null)
```

File: src/Core/Logger/Factory/AbstractLoggerTypeFactory.php [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+				return LogLevel::WARNING;
+			// legacy TRACE
+			case "2":
+				return LogLevel::NOTICE;
+			// legacy DEBUG
+			case "3":
+				return LogLevel::INFO;
+			// legacy DATA
+			case "4":
+				// legacy ALL
```
Comment (MrPetovan) [inline src/Core/Logger/Factory/AbstractLoggerTypeFactory.php]:
```suggestion
			// legacy ALL
```

File: src/Core/Logger/Factory/Logger.php [hunk 14]
```diff
@@ -22,134 +22,46 @@
...
 
-	/**
-	 * Creates a new PSR-3 compliant logger instances
-	 *
-	 * @param string|null $loglevel (optional) A given loglevel in case the loglevel in the config isn't applicable
-	 *
-	 * @return LoggerInterface The PSR-3 compliant logger instance
-	 */
-	public function create(string $loglevel = null): LoggerInterface
+	public function create(ICanCreateInstances $createInstances, IManageConfigValues $config, Profiler $profiler): LoggerInterface
```
Comment (MrPetovan) [inline src/Core/Logger/Factory/Logger.php]:
```suggestion
	public function create(ICanCreateInstances $instanceCreator, IManageConfigValues $config, Profiler $profiler): LoggerInterface
```

File: src/DI.php [hunk 15]
```diff
@@ -295,6 +297,11 @@ public static function flushLogger()
 		static::init($flushDice);
 	}
 
+	public static function loggCheck(): ICheckLoggerSettings
```
Comment (MrPetovan) [inline src/DI.php]:
```suggestion
	public static function logCheck(): ICheckLoggerSettings
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]
How do the hooks declared in this PR related to existing hooks?

**Comment by nupplaphil** [comment]
> How do the hooks declared in this PR related to existing hooks?

The existing hooks are all different kinds. This PR introduces the strategy hooks, which are used for dynamically load instance "strategies", like for
- Cache
- Storage
- Lock
- Logger

Currently, the hook returns an array and you need to extract the instance out of an array-index, like

```php
$data = [
  'instance' => null,
  'name' => 'aName',
];

Hook::callAll('instance_creation', $data);

if ($data['instance'] !== null) {
   $instance = $data['instance'];
}
```

now, it's called:
```php
$instance = $instanceRegister->create(InstanceInterface::class, 'aName');
```

and instead of 
```php
Hook::register('instance_creation', 'aName_instance_creation()`);
```

you need to set at `{addon}/static/hooks.config.php` or `static/hooks.config.php`:
```php
  HookType::STRATEGY => [
     InstanceInterface::class => [
        ANameInstance::class => ['aName'],
     ],
  ],
```
which automatically gets registered at runtime.

**Comment by MrPetovan** [comment]
I really would like to avoid the confusion between the existing data hooks and the new instance hooks. Can we just use the term "Strategies" instead of "Strategy Hooks"?

**Comment by nupplaphil** [comment]
In the end, all hooks are replaced by either strategies, pipelines or events. But yes, this naming convention would work as well. But they should stay inside the hooks namespace, shouldn't it?

Or do you have a different common namespace in mind?

I see this pattern as a possibility for decoupling the core logic as well :-).

Events for one-way events to all listeners, pipelines for dynamically process content and strategies to choose different implementations.

But that's just a possibility, for now they should replace the `Hook` logic and corresponding table.

**Comment by MrPetovan** [comment]
Ah, I see what you mean. In this case yes, the overarching Hook namespace still works. The transition might be a little confusing with a `hooks.config.php` that doesn't have anything to do with the current hooks.

**Comment by nupplaphil** [comment]
Shall I rename it? I'm open to any suggestions :D

**Comment by MrPetovan** [comment]
I'd rather we had a separate `strategies.config.php` file until we have all concepts under the same umbrella.

File: src/Core/Hooks/Capabilities/ICanRegisterInstances.php (hunk 6 above)
Comment (nupplaphil) [reply src/Core/Hooks/Capabilities/ICanRegisterInstances.php]: yep, you're right

File: src/Core/Hooks/Util/HookFileManager.php (hunk 8 above)
Comment (nupplaphil) [reply src/Core/Hooks/Util/HookFileManager.php]: moved the load-routine to the dependency file

**Comment by nupplaphil** [comment]
I'm finished here ;-)

**Review by MrPetovan — APPROVED** [review]

**Pull request merged**